## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [48]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [49]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [50]:
# Use the describe function to gather some basic statistics
df.describe

<bound method NDFrame.describe of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id

In [51]:
# Use this space to do any additional research
# and familiarize yourself with the data.
renamed_df = df.rename(columns={'first': 'First Name', 
                                'last': 'Last Name', 
                                'job': 'Job', 
                                'phone': 'Phone', 
                                'email': 'Email', 
                                'client_id' : 'Client ID', 
                                'order_id' : 'Order ID', 
                                'order_date' : 'Order Date', 
                                'order_week' : 'Order Week',
                                'order_year' : 'Order Year',
                                'item_id' : 'Item ID',
                                'category' : 'Category',
                                'subcategory' : 'Subcategory',
                                'unit_price' : 'Unit Price',
                                'unit_cost' : 'Unit Cost',
                                'unit_weight' : 'Unit Weight',
                                'qty' : 'Quantity',
                                'line_number' : 'Line Number'})
renamed_df.shape
renamed_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   First Name   54639 non-null  object 
 1   Last Name    54639 non-null  object 
 2   Job          54639 non-null  object 
 3   Phone        54639 non-null  object 
 4   Email        54639 non-null  object 
 5   Client ID    54639 non-null  int64  
 6   Order ID     54639 non-null  int64  
 7   Order Date   54639 non-null  object 
 8   Order Week   54639 non-null  int64  
 9   Order Year   54639 non-null  int64  
 10  Item ID      54639 non-null  object 
 11  Category     54639 non-null  object 
 12  Subcategory  54639 non-null  object 
 13  Unit Price   54639 non-null  float64
 14  Unit Cost    54639 non-null  float64
 15  Unit Weight  54639 non-null  float64
 16  Quantity     54639 non-null  int64  
 17  Line Number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [52]:
# What three item categories had the most entries?

categories_df = renamed_df["Category"].value_counts()
categories_df.head(3)



Category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [53]:
# For the category with the most entries, which subcategory had the most entries?

top_category = renamed_df["Category"].value_counts().index[0]
top_category_df = renamed_df[renamed_df['Category'] == top_category]
top_category_df["Subcategory"].value_counts().head(1)


Subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [54]:
# Which five clients had the most entries in the data?

renamed_df['Client ID'].value_counts().head(5)



Client ID
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [55]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = renamed_df['Client ID'].value_counts().head(5).index.to_list()
print(top_client_ids)


[33615, 66037, 46820, 38378, 24741]


In [56]:
# How many total units (the qty column) did the client with the most entries order order?
top_client = df['client_id'].value_counts().head(1).index[0]

top_client_df = renamed_df[renamed_df['Client ID'] == top_client]

total_units = top_client_df['Quantity'].sum()

print(f'The top client ordered {total_units} units.')

The top client ordered 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [57]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty

subtotal = renamed_df['Unit Price'] * renamed_df['Quantity']
renamed_df['Subtotal'] = subtotal

renamed_df.head()

First Name Last Name                    Job               Phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year         Item ID     Category         Subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   Unit Price  Unit Cost  Unit Weight  Quantity  Line Number   Subtotal  
0     1096.80     762.71         7.50       105            1  115164.00  
1       24.95      15.09         1.49        21            0     523.95  
2       13.52       7.86         1.68        39            6     527.28  
3       36.42      24.85         1.23        29            3    1056.18  
4      195.10     108.17        46.43        20            1    3902.00

In [58]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

renamed_df['Total Weight'] = renamed_df['Unit Weight'] * renamed_df['Quantity']

def shipping_price_calc (total_weight):
    if (total_weight < 50):
        return total_weight * 10
    else:
        return total_weight * 7
    
renamed_df['Shipping Price'] = renamed_df['Total Weight'].apply(shipping_price_calc)
renamed_df.head()
        



First Name Last Name                    Job               Phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year  ...     Category         Subcategory Unit Price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   Unit Cost  Unit Weight  Quantity  Line Number   Subtotal  Total Weight  \
0     762.71         7.50       105            1  115164.00        787.50   
1      15.09         1.49        21            0     523.95         31.29   
2       7.86         1.68        39            6     527.28         65.52   
3      24.85         1.23        29            3    1056.18         35.67   
4     108.17        46.43        20            1    3902.00        928.60   

   Shipping Price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [59]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
total_price = (renamed_df['Shipping Price'] + renamed_df['Subtotal']) * 1.0925
renamed_df['Line Price'] = round(total_price, 2)

renamed_df.head()

First Name Last Name                    Job               Phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year  ...         Subcategory Unit Price Unit Cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   Unit Weight  Quantity  Line Number   Subtotal  Total Weight  \
0         7.50       105            1  115164.00        787.50   
1         1.49        21            0     523.95         31.29   
2         1.68        39            6     527.28         65.52   
3         1.23        29            3    1056.18         35.67   
4        46.43        20            1    3902.00        928.60   

   Shipping Price  Line Price  
0         5512.50   131839.08  
1          312.90      914.26  
2          458.64     1077.12  
3          356.70     1543.57  
4         6500.20    11364.40  

[5 rows x 22 columns]

In [60]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
line_cost = (renamed_df['Unit Cost'] * renamed_df['Quantity']) + renamed_df['Shipping Price']
renamed_df['Line Cost'] = line_cost

renamed_df.head()


First Name Last Name                    Job               Phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year  ... Unit Price Unit Cost Unit Weight  Quantity  \
0          17        2023  ...    1096.80    762.71        7.50       105   
1          20        2023  ...      24.95     15.09        1.49        21   
2           4        2023  ...      13.52      7.86        1.68        39   
3          17        2023  ...      36.42     24.85        1.23        29   
4          21        2023  ...     195.10    108.17       46.43        20   

   Line Number   Subtotal  Total Weight  Shipping Price  Line Price  Line Cost  
0            1  115164.00        787.50         5512.50   131839.08   85597.05  
1            0     523.95         31.29          312.90      914.26     629.79  
2            6     527.28         65.52          458.64     1077.12     765.18  
3            3    1056.18         35.67          356.70     1543.57    1077.35  
4            1    3902.00        928.60         6500.20    11364.40    8663.60  

[5 rows x 23 columns]

In [61]:
# Create a column for the profit of each line using line cost and line price
line_profit = renamed_df['Line Price'] - renamed_df['Line Cost']
renamed_df['Profit'] = line_profit

renamed_df.head()


First Name Last Name                    Job               Phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year  ... Unit Cost Unit Weight Quantity  Line Number  \
0          17        2023  ...    762.71        7.50      105            1   
1          20        2023  ...     15.09        1.49       21            0   
2           4        2023  ...      7.86        1.68       39            6   
3          17        2023  ...     24.85        1.23       29            3   
4          21        2023  ...    108.17       46.43       20            1   

    Subtotal  Total Weight  Shipping Price  Line Price  Line Cost    Profit  
0  115164.00        787.50         5512.50   131839.08   85597.05  46242.03  
1     523.95         31.29          312.90      914.26     629.79    284.47  
2     527.28         65.52          458.64     1077.12     765.18    311.94  
3    1056.18         35.67          356.70     1543.57    1077.35    466.22  
4    3902.00        928.60         6500.20    11364.40    8663.60   2700.80  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [62]:
# Check your work using the totals above

def order_check (order_id):
    order_total = round(renamed_df[renamed_df['Order ID'] == order_id]['Line Price'].sum(), 2)
    print(f'The total price for the order is ${order_total}.')
    return

order_check(2742071)
order_check(2173913)
order_check(6128929)

The total price for the order is $152811.89.
The total price for the order is $162388.71.
The total price for the order is $923441.25.


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [63]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
def client_spending_total (client_id):
    spending_total = round(renamed_df[renamed_df['Client ID'] == client_id]['Line Price'].sum(), 2)
    print(client_id, spending_total)

for client in top_client_ids:
    client_spending_total(client)


33615 8377144.65
66037 10259514.79
46820 9743794.36
38378 12906550.87
24741 82268892.02


In [64]:
df_x = renamed_df[renamed_df['Client ID'] == 33615]
print(df_x[['Line Price']])

       Line Price
5        16392.81
12      262060.25
724       6684.55
840       1213.72
1405      1066.43
...           ...
52678     4454.08
53450     7146.74
53678     6600.44
53995    72005.40
54301      763.00

[220 rows x 1 columns]


In [65]:
# Create a summary DataFrame showing the totals for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_clients_df = renamed_df[renamed_df['Client ID'].isin(top_client_ids)]

top_client_summary_df = top_clients_df.groupby('Client ID').agg({
    'Quantity': 'sum',
    'Shipping Price': 'sum',
    'Line Price': 'sum',
    'Line Cost': 'sum',
    'Profit': 'sum'
}).reset_index()

print(top_client_summary_df)

   Client ID  Quantity  Shipping Price   Line Price    Line Cost       Profit
0      24741    239862      5126448.37  82268892.02  45688899.71  36579992.31
1      33615     64313      1828834.89   8377144.65   6175163.91   2201980.74
2      38378     73667      3429455.40  12906550.87   9634720.98   3271829.89
3      46820     75768      1601448.84   9743794.36   7007191.64   2736602.72
4      66037     43018      1395151.85  10259514.79   7004482.98   3255031.81


In [72]:
# Format the data and rename the columns to names suitable for presentation.
formated_top_client_df = top_client_summary_df.rename(columns={'Quantity' : 'Units',
                                                               'Shipping Price' : 'Shipping',
                                                               'Line Price' : 'Total Revenue',
                                                               'Line Cost' : 'Total Cost',
                                                               'Profit' : 'Total Profit'
                                                               })

# Define the money columns. 

# Define a function that converts a dollar amount to millions.
def currency_format_millions(amount):
    return round(amount/1000000, 6)

# Apply the currency_format_millions function to only the money columns. 
formated_top_client_df['Shipping'] = formated_top_client_df['Shipping'].apply(currency_format_millions)
formated_top_client_df['Total Revenue'] = formated_top_client_df['Total Revenue'].apply(currency_format_millions)
formated_top_client_df['Total Cost'] = formated_top_client_df['Total Cost'].apply(currency_format_millions)
formated_top_client_df['Total Profit'] = formated_top_client_df['Total Profit'].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
formated_top_client_df.rename(columns={'Shipping' : 'Shipping (millions)',
                                       'Total Revenue' : 'Total Revenue (millions)',
                                       'Total Cost' : 'Total Cost (millions)',
                                       'Total Profit' : 'Total Profit (millions)'
}, inplace=True)

display(formated_top_client_df)



Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      24741  239862             5.126448                 82.268892   
1      33615   64313             1.828835                  8.377145   
2      38378   73667             3.429455                 12.906551   
3      46820   75768             1.601449                  9.743794   
4      66037   43018             1.395152                 10.259515   

   Total Cost (millions)  Total Profit (millions)  
0              45.688900                36.579992  
1               6.175164                 2.201981  
2               9.634721                 3.271830  
3               7.007192                 2.736603  
4               7.004483                 3.255032

In [73]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DataFrame.
final_formated_data = formated_top_client_df.sort_values(by= 'Total Profit (millions)', ascending= False)
display(formated_top_client_df)

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      24741  239862             5.126448                 82.268892   
1      33615   64313             1.828835                  8.377145   
2      38378   73667             3.429455                 12.906551   
3      46820   75768             1.601449                  9.743794   
4      66037   43018             1.395152                 10.259515   

   Total Cost (millions)  Total Profit (millions)  
0              45.688900                36.579992  
1               6.175164                 2.201981  
2               9.634721                 3.271830  
3               7.007192                 2.736603  
4               7.004483                 3.255032

## Summary:

Using the pandas data library, I was able to import client_dataset.csv into a dataframe and use this dataframe to examine and enrich the data to create summary statistics by grouping by client ID. I was then able to format the dataset to create a more digestible and readable format to analyze the wholesale data. Theses tools were combined to determine the clients with the largest profits, most orders, and the most common item categories purchased.